src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+            {
+                return new StackFrameParser(text).TryParseTree();
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="VirtualCharSequence"/> and calls <see cref="TryParse(VirtualCharSequence)"/>
+        /// </summary>
+        public static StackFrameTree? TryParse(string text)
+            => TryParse(VirtualCharSequence.Create(0, text));
+
+        /// <summary>
+        /// Attempts to parse the full tree. Returns null on malformed data
+        /// </summary>
+        private StackFrameTree? TryParseTree()
+        {
+            var (_, methodDeclaration) = TryParseMethodDeclaration();
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            var fileInformationResult = TryParseFileInformation();
+            if (!fileInformationResult.Success)
+            {
+                return null;
+            }
+
+            var remainingTrivia = _lexer.TryScanRemainingTrivia();
+
+            var eolToken = CurrentCharAsToken().With(leadingTrivia: remainingTrivia.HasValue ? ImmutableArray.Create(remainingTrivia.Value) : ImmutableArray<StackFrameTrivia>.Empty);
+
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(eolToken.Kind == StackFrameKind.EndOfLine);
+
+            var root = new StackFrameCompilationUnit(methodDeclaration, fileInformationResult.Value, eolToken);
+
+            return new(_lexer.Text, root);
+        }
+
+        /// <summary>
+        /// Attempts to parse the full method declaration, optionally adding leading whitespace as trivia. Includes
+        /// all of the generic indicators for types, 
+        /// 
+        /// Ex: [|MyClass.MyMethod(string s)|]
+        /// </summary>
+        private ParseResult<StackFrameMethodDeclarationNode> TryParseMethodDeclaration()
+        {
+            var (success, identifierNode) = TryParseNameNode(scanAtTrivia: true);
+            if (!success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            if (identifierNode is not StackFrameQualifiedNameNode memberAccessExpression)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var typeArgumentsResult = TryParseTypeArguments();
+            if (!typeArgumentsResult.Success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var argumentsResult = ParseMethodParameters();
+            if (!argumentsResult.Success || argumentsResult.Value is null)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            return new(new(memberAccessExpression, typeArgumentsResult.Value, argumentsResult.Value));
```

Comment:
consider an implicit conversion from a *non-null* T to a success result.  but up to you.  that might be more confusing.

Reply:
I like this, especially because you can't use implicit new here so the type is still apparent.